packages/std/src/testing/mock.rs

```diff
@@ -142,62 +122,31 @@ impl Api for MockApi {
                 "Invalid input: address not normalized",
             ));
         }
-
         Ok(Addr::unchecked(input))
     }
 
     fn addr_canonicalize(&self, input: &str) -> StdResult<CanonicalAddr> {
-        // Dummy input validation. This is more sophisticated for formats like bech32, where format and checksum are validated.
-        let min_length = 3;
-        let max_length = self.canonical_length;
-        if input.len() < min_length {
-            return Err(StdError::generic_err(
-                format!("Invalid input: human address too short for this mock implementation (must be >= {min_length})."),
-            ));
-        }
-        if input.len() > max_length {
-            return Err(StdError::generic_err(
-                format!("Invalid input: human address too long for this mock implementation (must be <= {max_length})."),
-            ));
-        }
-
-        // mimics formats like hex or bech32 where different casings are valid for one address
-        let normalized = input.to_lowercase();
-
-        let mut out = Vec::from(normalized);
-
-        // pad to canonical length with NULL bytes
-        out.resize(self.canonical_length, 0x00);
-        // content-dependent rotate followed by shuffle to destroy
-        // the most obvious structure (https://github.com/CosmWasm/cosmwasm/issues/552)
-        let rotate_by = digit_sum(&out) % self.canonical_length;
-        out.rotate_left(rotate_by);
-        for _ in 0..SHUFFLES_ENCODE {
-            out = riffle_shuffle(&out);
+        if let Ok((prefix, decoded, Variant::Bech32)) = decode(input) {
+            if prefix == self.bech32_prefix {
+                if let Ok(bytes) = Vec::<u8>::from_base32(&decoded) {
+                    validate_length(&bytes)?;
+                    return Ok(bytes.into());
+                }
+            }
         }
-        Ok(out.into())
+        Err(StdError::generic_err("Invalid input"))
```

Comment:
Breaking this down into different error messages would be very helpful for contract devs. At least the 3 cases should be separated:
- Bech32 decoding error
- Wrong prefix
- invalid length